plugins/kubernetes-backend/src/service/KubernetesProxy.ts

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright 2022 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  ErrorResponseBody,
+  ForwardedError,
+  InputError,
+  NotFoundError,
+  serializeError,
+} from '@backstage/errors';
+import { bufferFromFileOrString } from '@kubernetes/client-node';
+import type { Request, RequestHandler } from 'express';
+import { createProxyMiddleware } from 'http-proxy-middleware';
+import { Logger } from 'winston';
+import { ClusterDetails, KubernetesClustersSupplier } from '../types/types';
+
+export const APPLICATION_JSON: string = 'application/json';
+
+/**
+ * The header that is used to specify the cluster name.
+ *
+ * @alpha
+ */
+export const HEADER_KUBERNETES_CLUSTER: string = 'X-Kubernetes-Cluster';
+
+/**
+ * A proxy that routes requests to the Kubernetes API.
+ *
+ * @alpha
+ */
+export class KubernetesProxy {
+  private readonly middlewareForClusterName = new Map<string, RequestHandler>();
+
+  constructor(
+    private readonly logger: Logger,
+    private readonly clusterSupplier: KubernetesClustersSupplier,
+  ) {}
+
+  public createRequestHandler(): RequestHandler {
+    return async (req, res, next) => {
+      const middleware = await this.getMiddleware(req);
+      middleware(req, res, next);
+    };
+  }
+
+  // We create one middleware per remote cluster and hold on to them, because
+  // the secure property isn't possible to decide on a per-request basis with a
+  // single middleware instance - and we don't expect it to change over time.
+  private async getMiddleware(originalReq: Request): Promise<RequestHandler> {
+    const originalCluster = await this.getClusterForRequest(originalReq);
+    let middleware = this.middlewareForClusterName.get(originalCluster.name);
+    if (!middleware) {
+      // Probably too risky without permissions protecting this endpoint
```

Comment:
+1 to this, we'd need to think a bit more about how to handle cluster auth if we want to be injecting credentials. In general, the set of permissions granted to the k8s plugin backend are likely rarely the same as the permissions you want to grant individual users.